{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7eb8a552-57db-4291-a470-a2f13a651000",
   "metadata": {},
   "source": [
    "# Summary\n",
    "\n",
    "Here, we show the process using scTSS to identify TSS and using Brie2 to detect alternative TSS in NPC dataset. \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "511d469d-9138-4879-ada4-9344b16825be",
   "metadata": {},
   "source": [
    "# Content\n",
    "\n",
    "* run scTSS\n",
    "  - merge bam file of NPC and NLH\n",
    "  - run scTSS\n",
    "* use Brie2 to detect alternative TSS usage\n",
    "  - make h5ad file\n",
    "  - get h5ad file of each cell type\n",
    "  - get cdr file\n",
    "  - run Brie2\n",
    "* do filtering (FDR <0.01)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33b0a1d8-b804-4519-bc14-1e3e6c6fa6c9",
   "metadata": {},
   "source": [
    "# run scTSS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6c47a03-a0e1-4371-a0ac-5cb03d1f0c72",
   "metadata": {},
   "source": [
    "## merge bam file of NPC and NLH"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a76bb023-464e-498e-91a1-21432a4bd890",
   "metadata": {},
   "source": [
    "nohup samtools merge  NPC_NLH.bam -b bamlist.fofn  -@ 20 --write-index &\n",
    "\n",
    "The bamlist.fofn looks like this:\n",
    "\n",
    "/storage/yhhuang/users/ruiyan/NPC/srafiledir/NLH/NLH.bam\n",
    "/storage/yhhuang/users/ruiyan/NPC/srafiledir/NPC/NPC.bam "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c87000a-6b67-4553-8169-f73ea112fd56",
   "metadata": {},
   "source": [
    "## run scTSS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "696dd1f0-de63-4faa-bb5f-21a558deca5f",
   "metadata": {},
   "source": [
    "#!/bin/bash\n",
    "\n",
    "gtfFile=/storage/yhhuang/users/ruiyan/annotation/refdata-gex-GRCh38-2020-A/genes/genes.gtf\n",
    "testgtfFile=/home/houruiyan/old_sctss/scTSS/test/Homo_sapiens.GRCh38.105.chr_test.gtf\n",
    "fastaFile=/storage/yhhuang/users/ruiyan/annotation/refdata-gex-GRCh38-2020-A/fasta/genome.fa\n",
    "\n",
    "\n",
    "scTSS-count --gtf $gtfFile --refFastq  $fastaFile  --bam /storage/yhhuang/users/ruiyan/NPC/srafiledir/scTSS_rerun/rawdata/NPC_NLH.bam  -c /storage/yhhuang/users/ruiyan/NPC/cdr_file/all_cell.tsv  -o /storage/yhhuang/users/ruiyan/NPC/srafiledir/scTSS_out/all_unannotation  --mode Unannotation --minCount 50  -p 20  --maxReadCount 10000 --clusterDistance 500"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4d3b390-cad0-4d0f-9848-6b269b15ed0b",
   "metadata": {},
   "source": [
    "# use Brie2 to detect alternative TSS usage"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57ab0057-ce7a-48f1-8030-0a568095dd11",
   "metadata": {},
   "source": [
    "## prepare file to run Brie2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8facae7-1657-475e-b930-989c8c01c92c",
   "metadata": {},
   "source": [
    "### make h5ad file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "678028a2-ab18-4956-91ad-9f8abaefe00e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a084046a-59c0-40c3-a36f-7813e63bd566",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f94c26f3-8cc7-4316-9ca5-dc51879f776e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 37559 × 5449\n",
       "    obs: 'cell_id', 'Anno_level_1', 'Anno_level_2', 'Anno_level_3', 'Anno_level_4', 'Anno_level_5', 'Anno_level_fig1', 'Sample', 'donor', 'organ', 'sort', 'file', 'Anno_stage', 'Age', 'Gender', 'Source'\n",
       "    var: 'TSS_start', 'TSS_end', 'gene_id', 'Chromosome', 'Feature', 'Start', 'End', 'Strand', 'gene_name', 'len'\n",
       "    uns: 'log1p'\n",
       "    obsm: 'X_umap'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata=sc.read('/storage3/yhhuang/users/ruiyan/development_data/processed_file/rerun_all_preprocessing_alllen_norm.h5ad')\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "77550e9c-79d7-4de5-b445-d681121445bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 5399108 × 36601\n",
       "    obs: 'index', 'cell_id', 'batch'\n",
       "    var: 'gene_ids', 'feature_types'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expadata=sc.read('/storage3/yhhuang/users/ruiyan/development_data/processed_file/cellranger_changeCellID/all_cellranger_development.h5ad')\n",
    "expadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "15c46c14-8f28-495a-bede-9a8452490658",
   "metadata": {},
   "outputs": [],
   "source": [
    "expadata.var['gene_name']=expadata.var.index\n",
    "expadata.var.index=expadata.var['gene_ids']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2ee01cb3-b69c-413c-9ce6-5248e6fcb935",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_ids</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>FAM138A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AL627309.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239945</th>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AL627309.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277836</th>\n",
       "      <td>ENSG00000277836</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AC141272.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278633</th>\n",
       "      <td>ENSG00000278633</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AC023491.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000276017</th>\n",
       "      <td>ENSG00000276017</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AC007325.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278817</th>\n",
       "      <td>ENSG00000278817</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AC007325.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277196</th>\n",
       "      <td>ENSG00000277196</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AC007325.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36601 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gene_ids    feature_types    gene_name\n",
       "gene_ids                                                      \n",
       "ENSG00000243485  ENSG00000243485  Gene Expression  MIR1302-2HG\n",
       "ENSG00000237613  ENSG00000237613  Gene Expression      FAM138A\n",
       "ENSG00000186092  ENSG00000186092  Gene Expression        OR4F5\n",
       "ENSG00000238009  ENSG00000238009  Gene Expression   AL627309.1\n",
       "ENSG00000239945  ENSG00000239945  Gene Expression   AL627309.3\n",
       "...                          ...              ...          ...\n",
       "ENSG00000277836  ENSG00000277836  Gene Expression   AC141272.1\n",
       "ENSG00000278633  ENSG00000278633  Gene Expression   AC023491.2\n",
       "ENSG00000276017  ENSG00000276017  Gene Expression   AC007325.1\n",
       "ENSG00000278817  ENSG00000278817  Gene Expression   AC007325.4\n",
       "ENSG00000277196  ENSG00000277196  Gene Expression   AC007325.2\n",
       "\n",
       "[36601 rows x 3 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expadata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f9c2ed9c-e993-41ff-86ea-136093047ea8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 5399108 × 36601\n",
       "    obs: 'index', 'cell_id', 'batch'\n",
       "    var: 'gene_ids', 'feature_types', 'gene_name'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expadata.obs_names_make_unique()\n",
    "\n",
    "expadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "adec6cf4-a338-4e63-ade9-ce2fbb19595b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TSS_start</th>\n",
       "      <th>TSS_end</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>len</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000197530_newTSS</th>\n",
       "      <td>1629513</td>\n",
       "      <td>1629602</td>\n",
       "      <td>ENSG00000197530</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>1615414</td>\n",
       "      <td>1630610</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>15196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000197530_ENST00000487053</th>\n",
       "      <td>1615842</td>\n",
       "      <td>1615943</td>\n",
       "      <td>ENSG00000197530</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>1615414</td>\n",
       "      <td>1630610</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>15196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000197530_ENST00000518681</th>\n",
       "      <td>1615420</td>\n",
       "      <td>1615554</td>\n",
       "      <td>ENSG00000197530</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>1615414</td>\n",
       "      <td>1630610</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>15196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000197530_ENST00000510793</th>\n",
       "      <td>1616193</td>\n",
       "      <td>1616435</td>\n",
       "      <td>ENSG00000197530</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>1615414</td>\n",
       "      <td>1630610</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>15196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000189409_newTSS</th>\n",
       "      <td>1632164</td>\n",
       "      <td>1632227</td>\n",
       "      <td>ENSG00000189409</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>1632162</td>\n",
       "      <td>1635263</td>\n",
       "      <td>+</td>\n",
       "      <td>MMP23B</td>\n",
       "      <td>3101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000130830_newTSS</th>\n",
       "      <td>154799743</td>\n",
       "      <td>154799794</td>\n",
       "      <td>ENSG00000130830</td>\n",
       "      <td>X</td>\n",
       "      <td>gene</td>\n",
       "      <td>154778683</td>\n",
       "      <td>154821007</td>\n",
       "      <td>-</td>\n",
       "      <td>MPP1</td>\n",
       "      <td>42324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000067048_ENST00000493363</th>\n",
       "      <td>12904785</td>\n",
       "      <td>12904950</td>\n",
       "      <td>ENSG00000067048</td>\n",
       "      <td>Y</td>\n",
       "      <td>gene</td>\n",
       "      <td>12904107</td>\n",
       "      <td>12920478</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX3Y</td>\n",
       "      <td>16371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000067048_ENST00000440554</th>\n",
       "      <td>12905701</td>\n",
       "      <td>12905772</td>\n",
       "      <td>ENSG00000067048</td>\n",
       "      <td>Y</td>\n",
       "      <td>gene</td>\n",
       "      <td>12904107</td>\n",
       "      <td>12920478</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX3Y</td>\n",
       "      <td>16371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000183878_ENST00000479713</th>\n",
       "      <td>13479800</td>\n",
       "      <td>13479952</td>\n",
       "      <td>ENSG00000183878</td>\n",
       "      <td>Y</td>\n",
       "      <td>gene</td>\n",
       "      <td>13248378</td>\n",
       "      <td>13480673</td>\n",
       "      <td>-</td>\n",
       "      <td>UTY</td>\n",
       "      <td>232295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000183878_newTSS</th>\n",
       "      <td>13479706</td>\n",
       "      <td>13479797</td>\n",
       "      <td>ENSG00000183878</td>\n",
       "      <td>Y</td>\n",
       "      <td>gene</td>\n",
       "      <td>13248378</td>\n",
       "      <td>13480673</td>\n",
       "      <td>-</td>\n",
       "      <td>UTY</td>\n",
       "      <td>232295</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5449 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 TSS_start    TSS_end          gene_id  \\\n",
       "transcript_id                                                            \n",
       "ENSG00000197530_newTSS             1629513    1629602  ENSG00000197530   \n",
       "ENSG00000197530_ENST00000487053    1615842    1615943  ENSG00000197530   \n",
       "ENSG00000197530_ENST00000518681    1615420    1615554  ENSG00000197530   \n",
       "ENSG00000197530_ENST00000510793    1616193    1616435  ENSG00000197530   \n",
       "ENSG00000189409_newTSS             1632164    1632227  ENSG00000189409   \n",
       "...                                    ...        ...              ...   \n",
       "ENSG00000130830_newTSS           154799743  154799794  ENSG00000130830   \n",
       "ENSG00000067048_ENST00000493363   12904785   12904950  ENSG00000067048   \n",
       "ENSG00000067048_ENST00000440554   12905701   12905772  ENSG00000067048   \n",
       "ENSG00000183878_ENST00000479713   13479800   13479952  ENSG00000183878   \n",
       "ENSG00000183878_newTSS            13479706   13479797  ENSG00000183878   \n",
       "\n",
       "                                Chromosome Feature      Start        End  \\\n",
       "transcript_id                                                              \n",
       "ENSG00000197530_newTSS                   1    gene    1615414    1630610   \n",
       "ENSG00000197530_ENST00000487053          1    gene    1615414    1630610   \n",
       "ENSG00000197530_ENST00000518681          1    gene    1615414    1630610   \n",
       "ENSG00000197530_ENST00000510793          1    gene    1615414    1630610   \n",
       "ENSG00000189409_newTSS                   1    gene    1632162    1635263   \n",
       "...                                    ...     ...        ...        ...   \n",
       "ENSG00000130830_newTSS                   X    gene  154778683  154821007   \n",
       "ENSG00000067048_ENST00000493363          Y    gene   12904107   12920478   \n",
       "ENSG00000067048_ENST00000440554          Y    gene   12904107   12920478   \n",
       "ENSG00000183878_ENST00000479713          Y    gene   13248378   13480673   \n",
       "ENSG00000183878_newTSS                   Y    gene   13248378   13480673   \n",
       "\n",
       "                                Strand gene_name     len  \n",
       "transcript_id                                             \n",
       "ENSG00000197530_newTSS               +      MIB2   15196  \n",
       "ENSG00000197530_ENST00000487053      +      MIB2   15196  \n",
       "ENSG00000197530_ENST00000518681      +      MIB2   15196  \n",
       "ENSG00000197530_ENST00000510793      +      MIB2   15196  \n",
       "ENSG00000189409_newTSS               +    MMP23B    3101  \n",
       "...                                ...       ...     ...  \n",
       "ENSG00000130830_newTSS               -      MPP1   42324  \n",
       "ENSG00000067048_ENST00000493363      +     DDX3Y   16371  \n",
       "ENSG00000067048_ENST00000440554      +     DDX3Y   16371  \n",
       "ENSG00000183878_ENST00000479713      -       UTY  232295  \n",
       "ENSG00000183878_newTSS               -       UTY  232295  \n",
       "\n",
       "[5449 rows x 10 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genedf=adata.var.copy()\n",
    "genedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ebe329c9-2558-4c22-9122-be99f14a0598",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TSS_start</th>\n",
       "      <th>TSS_end</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>len</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000197530_newTSS</th>\n",
       "      <td>1629513</td>\n",
       "      <td>1629602</td>\n",
       "      <td>ENSG00000197530</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>1615414</td>\n",
       "      <td>1630610</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>15196</td>\n",
       "      <td>256.473970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000197530_ENST00000487053</th>\n",
       "      <td>1615842</td>\n",
       "      <td>1615943</td>\n",
       "      <td>ENSG00000197530</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>1615414</td>\n",
       "      <td>1630610</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>15196</td>\n",
       "      <td>4405.169038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000197530_ENST00000518681</th>\n",
       "      <td>1615420</td>\n",
       "      <td>1615554</td>\n",
       "      <td>ENSG00000197530</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>1615414</td>\n",
       "      <td>1630610</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>15196</td>\n",
       "      <td>2676.125020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000197530_ENST00000510793</th>\n",
       "      <td>1616193</td>\n",
       "      <td>1616435</td>\n",
       "      <td>ENSG00000197530</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>1615414</td>\n",
       "      <td>1630610</td>\n",
       "      <td>+</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>15196</td>\n",
       "      <td>412.545578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000189409_newTSS</th>\n",
       "      <td>1632164</td>\n",
       "      <td>1632227</td>\n",
       "      <td>ENSG00000189409</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>1632162</td>\n",
       "      <td>1635263</td>\n",
       "      <td>+</td>\n",
       "      <td>MMP23B</td>\n",
       "      <td>3101</td>\n",
       "      <td>1120.169594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000130830_newTSS</th>\n",
       "      <td>154799743</td>\n",
       "      <td>154799794</td>\n",
       "      <td>ENSG00000130830</td>\n",
       "      <td>X</td>\n",
       "      <td>gene</td>\n",
       "      <td>154778683</td>\n",
       "      <td>154821007</td>\n",
       "      <td>-</td>\n",
       "      <td>MPP1</td>\n",
       "      <td>42324</td>\n",
       "      <td>1864.098669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000067048_ENST00000493363</th>\n",
       "      <td>12904785</td>\n",
       "      <td>12904950</td>\n",
       "      <td>ENSG00000067048</td>\n",
       "      <td>Y</td>\n",
       "      <td>gene</td>\n",
       "      <td>12904107</td>\n",
       "      <td>12920478</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX3Y</td>\n",
       "      <td>16371</td>\n",
       "      <td>21090.087880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000067048_ENST00000440554</th>\n",
       "      <td>12905701</td>\n",
       "      <td>12905772</td>\n",
       "      <td>ENSG00000067048</td>\n",
       "      <td>Y</td>\n",
       "      <td>gene</td>\n",
       "      <td>12904107</td>\n",
       "      <td>12920478</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX3Y</td>\n",
       "      <td>16371</td>\n",
       "      <td>916.549825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000183878_ENST00000479713</th>\n",
       "      <td>13479800</td>\n",
       "      <td>13479952</td>\n",
       "      <td>ENSG00000183878</td>\n",
       "      <td>Y</td>\n",
       "      <td>gene</td>\n",
       "      <td>13248378</td>\n",
       "      <td>13480673</td>\n",
       "      <td>-</td>\n",
       "      <td>UTY</td>\n",
       "      <td>232295</td>\n",
       "      <td>31537.631923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000183878_newTSS</th>\n",
       "      <td>13479706</td>\n",
       "      <td>13479797</td>\n",
       "      <td>ENSG00000183878</td>\n",
       "      <td>Y</td>\n",
       "      <td>gene</td>\n",
       "      <td>13248378</td>\n",
       "      <td>13480673</td>\n",
       "      <td>-</td>\n",
       "      <td>UTY</td>\n",
       "      <td>232295</td>\n",
       "      <td>442.788557</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5449 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 TSS_start    TSS_end          gene_id  \\\n",
       "transcript_id                                                            \n",
       "ENSG00000197530_newTSS             1629513    1629602  ENSG00000197530   \n",
       "ENSG00000197530_ENST00000487053    1615842    1615943  ENSG00000197530   \n",
       "ENSG00000197530_ENST00000518681    1615420    1615554  ENSG00000197530   \n",
       "ENSG00000197530_ENST00000510793    1616193    1616435  ENSG00000197530   \n",
       "ENSG00000189409_newTSS             1632164    1632227  ENSG00000189409   \n",
       "...                                    ...        ...              ...   \n",
       "ENSG00000130830_newTSS           154799743  154799794  ENSG00000130830   \n",
       "ENSG00000067048_ENST00000493363   12904785   12904950  ENSG00000067048   \n",
       "ENSG00000067048_ENST00000440554   12905701   12905772  ENSG00000067048   \n",
       "ENSG00000183878_ENST00000479713   13479800   13479952  ENSG00000183878   \n",
       "ENSG00000183878_newTSS            13479706   13479797  ENSG00000183878   \n",
       "\n",
       "                                Chromosome Feature      Start        End  \\\n",
       "transcript_id                                                              \n",
       "ENSG00000197530_newTSS                   1    gene    1615414    1630610   \n",
       "ENSG00000197530_ENST00000487053          1    gene    1615414    1630610   \n",
       "ENSG00000197530_ENST00000518681          1    gene    1615414    1630610   \n",
       "ENSG00000197530_ENST00000510793          1    gene    1615414    1630610   \n",
       "ENSG00000189409_newTSS                   1    gene    1632162    1635263   \n",
       "...                                    ...     ...        ...        ...   \n",
       "ENSG00000130830_newTSS                   X    gene  154778683  154821007   \n",
       "ENSG00000067048_ENST00000493363          Y    gene   12904107   12920478   \n",
       "ENSG00000067048_ENST00000440554          Y    gene   12904107   12920478   \n",
       "ENSG00000183878_ENST00000479713          Y    gene   13248378   13480673   \n",
       "ENSG00000183878_newTSS                   Y    gene   13248378   13480673   \n",
       "\n",
       "                                Strand gene_name     len         count  \n",
       "transcript_id                                                           \n",
       "ENSG00000197530_newTSS               +      MIB2   15196    256.473970  \n",
       "ENSG00000197530_ENST00000487053      +      MIB2   15196   4405.169038  \n",
       "ENSG00000197530_ENST00000518681      +      MIB2   15196   2676.125020  \n",
       "ENSG00000197530_ENST00000510793      +      MIB2   15196    412.545578  \n",
       "ENSG00000189409_newTSS               +    MMP23B    3101   1120.169594  \n",
       "...                                ...       ...     ...           ...  \n",
       "ENSG00000130830_newTSS               -      MPP1   42324   1864.098669  \n",
       "ENSG00000067048_ENST00000493363      +     DDX3Y   16371  21090.087880  \n",
       "ENSG00000067048_ENST00000440554      +     DDX3Y   16371    916.549825  \n",
       "ENSG00000183878_ENST00000479713      -       UTY  232295  31537.631923  \n",
       "ENSG00000183878_newTSS               -       UTY  232295    442.788557  \n",
       "\n",
       "[5449 rows x 11 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genedf['count']=np.sum(adata.X,axis=0)\n",
    "genedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5c16746f-e8ba-44ea-8d91-aa8f34a67045",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TSS_start</th>\n",
       "      <th>TSS_end</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>len</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460_ENST00000472795</th>\n",
       "      <td>169794679</td>\n",
       "      <td>169794816</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>169662006</td>\n",
       "      <td>169854080</td>\n",
       "      <td>+</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>192074</td>\n",
       "      <td>511.390458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460_ENST00000359326</th>\n",
       "      <td>169794982</td>\n",
       "      <td>169795207</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>169662006</td>\n",
       "      <td>169854080</td>\n",
       "      <td>+</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>192074</td>\n",
       "      <td>21252.589071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938_newTSS</th>\n",
       "      <td>27635064</td>\n",
       "      <td>27635076</td>\n",
       "      <td>ENSG00000000938</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>27612063</td>\n",
       "      <td>27635185</td>\n",
       "      <td>-</td>\n",
       "      <td>FGR</td>\n",
       "      <td>23122</td>\n",
       "      <td>201.604448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938_ENST00000457296</th>\n",
       "      <td>27626061</td>\n",
       "      <td>27626162</td>\n",
       "      <td>ENSG00000000938</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>27612063</td>\n",
       "      <td>27635185</td>\n",
       "      <td>-</td>\n",
       "      <td>FGR</td>\n",
       "      <td>23122</td>\n",
       "      <td>618.203238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001626_ENST00000647720</th>\n",
       "      <td>117592386</td>\n",
       "      <td>117592554</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>7</td>\n",
       "      <td>gene</td>\n",
       "      <td>117465783</td>\n",
       "      <td>117715971</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>250188</td>\n",
       "      <td>328.055856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285646_ENST00000650448</th>\n",
       "      <td>11908143</td>\n",
       "      <td>11908221</td>\n",
       "      <td>ENSG00000285646</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>11907939</td>\n",
       "      <td>11914298</td>\n",
       "      <td>+</td>\n",
       "      <td>AL021155.5</td>\n",
       "      <td>6359</td>\n",
       "      <td>13139.479124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285756_ENST00000671583</th>\n",
       "      <td>3843689</td>\n",
       "      <td>3843850</td>\n",
       "      <td>ENSG00000285756</td>\n",
       "      <td>X</td>\n",
       "      <td>gene</td>\n",
       "      <td>3817527</td>\n",
       "      <td>3937855</td>\n",
       "      <td>-</td>\n",
       "      <td>BX890604.2</td>\n",
       "      <td>120328</td>\n",
       "      <td>4975.422838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285756_newTSS</th>\n",
       "      <td>3882165</td>\n",
       "      <td>3882287</td>\n",
       "      <td>ENSG00000285756</td>\n",
       "      <td>X</td>\n",
       "      <td>gene</td>\n",
       "      <td>3817527</td>\n",
       "      <td>3937855</td>\n",
       "      <td>-</td>\n",
       "      <td>BX890604.2</td>\n",
       "      <td>120328</td>\n",
       "      <td>6140.318460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000286340_newTSS</th>\n",
       "      <td>79078243</td>\n",
       "      <td>79078427</td>\n",
       "      <td>ENSG00000286340</td>\n",
       "      <td>6</td>\n",
       "      <td>gene</td>\n",
       "      <td>79077840</td>\n",
       "      <td>79081371</td>\n",
       "      <td>+</td>\n",
       "      <td>AL356776.2</td>\n",
       "      <td>3531</td>\n",
       "      <td>299.303719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000286340_ENST00000669651</th>\n",
       "      <td>79078545</td>\n",
       "      <td>79078653</td>\n",
       "      <td>ENSG00000286340</td>\n",
       "      <td>6</td>\n",
       "      <td>gene</td>\n",
       "      <td>79077840</td>\n",
       "      <td>79081371</td>\n",
       "      <td>+</td>\n",
       "      <td>AL356776.2</td>\n",
       "      <td>3531</td>\n",
       "      <td>1454.454024</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4712 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 TSS_start    TSS_end          gene_id  \\\n",
       "transcript_id                                                            \n",
       "ENSG00000000460_ENST00000472795  169794679  169794816  ENSG00000000460   \n",
       "ENSG00000000460_ENST00000359326  169794982  169795207  ENSG00000000460   \n",
       "ENSG00000000938_newTSS            27635064   27635076  ENSG00000000938   \n",
       "ENSG00000000938_ENST00000457296   27626061   27626162  ENSG00000000938   \n",
       "ENSG00000001626_ENST00000647720  117592386  117592554  ENSG00000001626   \n",
       "...                                    ...        ...              ...   \n",
       "ENSG00000285646_ENST00000650448   11908143   11908221  ENSG00000285646   \n",
       "ENSG00000285756_ENST00000671583    3843689    3843850  ENSG00000285756   \n",
       "ENSG00000285756_newTSS             3882165    3882287  ENSG00000285756   \n",
       "ENSG00000286340_newTSS            79078243   79078427  ENSG00000286340   \n",
       "ENSG00000286340_ENST00000669651   79078545   79078653  ENSG00000286340   \n",
       "\n",
       "                                Chromosome Feature      Start        End  \\\n",
       "transcript_id                                                              \n",
       "ENSG00000000460_ENST00000472795          1    gene  169662006  169854080   \n",
       "ENSG00000000460_ENST00000359326          1    gene  169662006  169854080   \n",
       "ENSG00000000938_newTSS                   1    gene   27612063   27635185   \n",
       "ENSG00000000938_ENST00000457296          1    gene   27612063   27635185   \n",
       "ENSG00000001626_ENST00000647720          7    gene  117465783  117715971   \n",
       "...                                    ...     ...        ...        ...   \n",
       "ENSG00000285646_ENST00000650448          1    gene   11907939   11914298   \n",
       "ENSG00000285756_ENST00000671583          X    gene    3817527    3937855   \n",
       "ENSG00000285756_newTSS                   X    gene    3817527    3937855   \n",
       "ENSG00000286340_newTSS                   6    gene   79077840   79081371   \n",
       "ENSG00000286340_ENST00000669651          6    gene   79077840   79081371   \n",
       "\n",
       "                                Strand   gene_name     len         count  \n",
       "transcript_id                                                             \n",
       "ENSG00000000460_ENST00000472795      +    C1orf112  192074    511.390458  \n",
       "ENSG00000000460_ENST00000359326      +    C1orf112  192074  21252.589071  \n",
       "ENSG00000000938_newTSS               -         FGR   23122    201.604448  \n",
       "ENSG00000000938_ENST00000457296      -         FGR   23122    618.203238  \n",
       "ENSG00000001626_ENST00000647720      +        CFTR  250188    328.055856  \n",
       "...                                ...         ...     ...           ...  \n",
       "ENSG00000285646_ENST00000650448      +  AL021155.5    6359  13139.479124  \n",
       "ENSG00000285756_ENST00000671583      -  BX890604.2  120328   4975.422838  \n",
       "ENSG00000285756_newTSS               -  BX890604.2  120328   6140.318460  \n",
       "ENSG00000286340_newTSS               +  AL356776.2    3531    299.303719  \n",
       "ENSG00000286340_ENST00000669651      +  AL356776.2    3531   1454.454024  \n",
       "\n",
       "[4712 rows x 11 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keepdf=genedf.sort_values(['gene_id','count']).groupby('gene_id').tail(2)\n",
    "keepdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "04dcd98f-1212-444e-95b5-180ec2f3919d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TSS_start</th>\n",
       "      <th>TSS_end</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>len</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460_ENST00000472795</th>\n",
       "      <td>169794679</td>\n",
       "      <td>169794816</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>169662006</td>\n",
       "      <td>169854080</td>\n",
       "      <td>+</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>192074</td>\n",
       "      <td>511.390458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938_newTSS</th>\n",
       "      <td>27635064</td>\n",
       "      <td>27635076</td>\n",
       "      <td>ENSG00000000938</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>27612063</td>\n",
       "      <td>27635185</td>\n",
       "      <td>-</td>\n",
       "      <td>FGR</td>\n",
       "      <td>23122</td>\n",
       "      <td>201.604448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001626_ENST00000647720</th>\n",
       "      <td>117592386</td>\n",
       "      <td>117592554</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>7</td>\n",
       "      <td>gene</td>\n",
       "      <td>117465783</td>\n",
       "      <td>117715971</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>250188</td>\n",
       "      <td>328.055856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001631_newTSS</th>\n",
       "      <td>92245432</td>\n",
       "      <td>92245492</td>\n",
       "      <td>ENSG00000001631</td>\n",
       "      <td>7</td>\n",
       "      <td>gene</td>\n",
       "      <td>92198968</td>\n",
       "      <td>92246166</td>\n",
       "      <td>-</td>\n",
       "      <td>KRIT1</td>\n",
       "      <td>47198</td>\n",
       "      <td>1285.028807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000002834_newTSS</th>\n",
       "      <td>38870152</td>\n",
       "      <td>38870251</td>\n",
       "      <td>ENSG00000002834</td>\n",
       "      <td>17</td>\n",
       "      <td>gene</td>\n",
       "      <td>38869858</td>\n",
       "      <td>38921770</td>\n",
       "      <td>+</td>\n",
       "      <td>LASP1</td>\n",
       "      <td>51912</td>\n",
       "      <td>269.558158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285043_ENST00000395248</th>\n",
       "      <td>30053091</td>\n",
       "      <td>30053222</td>\n",
       "      <td>ENSG00000285043</td>\n",
       "      <td>16</td>\n",
       "      <td>gene</td>\n",
       "      <td>30053089</td>\n",
       "      <td>30070420</td>\n",
       "      <td>+</td>\n",
       "      <td>AC093512.2</td>\n",
       "      <td>17331</td>\n",
       "      <td>1403.487265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285331_newTSS</th>\n",
       "      <td>56886908</td>\n",
       "      <td>56887106</td>\n",
       "      <td>ENSG00000285331</td>\n",
       "      <td>15</td>\n",
       "      <td>gene</td>\n",
       "      <td>56795459</td>\n",
       "      <td>56918499</td>\n",
       "      <td>-</td>\n",
       "      <td>AC090517.5</td>\n",
       "      <td>123040</td>\n",
       "      <td>612.365203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285646_newTSS</th>\n",
       "      <td>11909037</td>\n",
       "      <td>11909155</td>\n",
       "      <td>ENSG00000285646</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>11907939</td>\n",
       "      <td>11914298</td>\n",
       "      <td>+</td>\n",
       "      <td>AL021155.5</td>\n",
       "      <td>6359</td>\n",
       "      <td>575.438952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285756_ENST00000671583</th>\n",
       "      <td>3843689</td>\n",
       "      <td>3843850</td>\n",
       "      <td>ENSG00000285756</td>\n",
       "      <td>X</td>\n",
       "      <td>gene</td>\n",
       "      <td>3817527</td>\n",
       "      <td>3937855</td>\n",
       "      <td>-</td>\n",
       "      <td>BX890604.2</td>\n",
       "      <td>120328</td>\n",
       "      <td>4975.422838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000286340_newTSS</th>\n",
       "      <td>79078243</td>\n",
       "      <td>79078427</td>\n",
       "      <td>ENSG00000286340</td>\n",
       "      <td>6</td>\n",
       "      <td>gene</td>\n",
       "      <td>79077840</td>\n",
       "      <td>79081371</td>\n",
       "      <td>+</td>\n",
       "      <td>AL356776.2</td>\n",
       "      <td>3531</td>\n",
       "      <td>299.303719</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2356 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 TSS_start    TSS_end          gene_id  \\\n",
       "transcript_id                                                            \n",
       "ENSG00000000460_ENST00000472795  169794679  169794816  ENSG00000000460   \n",
       "ENSG00000000938_newTSS            27635064   27635076  ENSG00000000938   \n",
       "ENSG00000001626_ENST00000647720  117592386  117592554  ENSG00000001626   \n",
       "ENSG00000001631_newTSS            92245432   92245492  ENSG00000001631   \n",
       "ENSG00000002834_newTSS            38870152   38870251  ENSG00000002834   \n",
       "...                                    ...        ...              ...   \n",
       "ENSG00000285043_ENST00000395248   30053091   30053222  ENSG00000285043   \n",
       "ENSG00000285331_newTSS            56886908   56887106  ENSG00000285331   \n",
       "ENSG00000285646_newTSS            11909037   11909155  ENSG00000285646   \n",
       "ENSG00000285756_ENST00000671583    3843689    3843850  ENSG00000285756   \n",
       "ENSG00000286340_newTSS            79078243   79078427  ENSG00000286340   \n",
       "\n",
       "                                Chromosome Feature      Start        End  \\\n",
       "transcript_id                                                              \n",
       "ENSG00000000460_ENST00000472795          1    gene  169662006  169854080   \n",
       "ENSG00000000938_newTSS                   1    gene   27612063   27635185   \n",
       "ENSG00000001626_ENST00000647720          7    gene  117465783  117715971   \n",
       "ENSG00000001631_newTSS                   7    gene   92198968   92246166   \n",
       "ENSG00000002834_newTSS                  17    gene   38869858   38921770   \n",
       "...                                    ...     ...        ...        ...   \n",
       "ENSG00000285043_ENST00000395248         16    gene   30053089   30070420   \n",
       "ENSG00000285331_newTSS                  15    gene   56795459   56918499   \n",
       "ENSG00000285646_newTSS                   1    gene   11907939   11914298   \n",
       "ENSG00000285756_ENST00000671583          X    gene    3817527    3937855   \n",
       "ENSG00000286340_newTSS                   6    gene   79077840   79081371   \n",
       "\n",
       "                                Strand   gene_name     len        count  \n",
       "transcript_id                                                            \n",
       "ENSG00000000460_ENST00000472795      +    C1orf112  192074   511.390458  \n",
       "ENSG00000000938_newTSS               -         FGR   23122   201.604448  \n",
       "ENSG00000001626_ENST00000647720      +        CFTR  250188   328.055856  \n",
       "ENSG00000001631_newTSS               -       KRIT1   47198  1285.028807  \n",
       "ENSG00000002834_newTSS               +       LASP1   51912   269.558158  \n",
       "...                                ...         ...     ...          ...  \n",
       "ENSG00000285043_ENST00000395248      +  AC093512.2   17331  1403.487265  \n",
       "ENSG00000285331_newTSS               -  AC090517.5  123040   612.365203  \n",
       "ENSG00000285646_newTSS               +  AL021155.5    6359   575.438952  \n",
       "ENSG00000285756_ENST00000671583      -  BX890604.2  120328  4975.422838  \n",
       "ENSG00000286340_newTSS               +  AL356776.2    3531   299.303719  \n",
       "\n",
       "[2356 rows x 11 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "firstdf=keepdf[::2]\n",
    "firstdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "923441a2-e79d-4fac-9deb-9c3aa7a184c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TSS_start</th>\n",
       "      <th>TSS_end</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>len</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460_ENST00000359326</th>\n",
       "      <td>169794982</td>\n",
       "      <td>169795207</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>169662006</td>\n",
       "      <td>169854080</td>\n",
       "      <td>+</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>192074</td>\n",
       "      <td>21252.589071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938_ENST00000457296</th>\n",
       "      <td>27626061</td>\n",
       "      <td>27626162</td>\n",
       "      <td>ENSG00000000938</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>27612063</td>\n",
       "      <td>27635185</td>\n",
       "      <td>-</td>\n",
       "      <td>FGR</td>\n",
       "      <td>23122</td>\n",
       "      <td>618.203238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001626_newTSS</th>\n",
       "      <td>117479966</td>\n",
       "      <td>117480057</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>7</td>\n",
       "      <td>gene</td>\n",
       "      <td>117465783</td>\n",
       "      <td>117715971</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>250188</td>\n",
       "      <td>796.037284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001631_ENST00000412043</th>\n",
       "      <td>92245654</td>\n",
       "      <td>92245871</td>\n",
       "      <td>ENSG00000001631</td>\n",
       "      <td>7</td>\n",
       "      <td>gene</td>\n",
       "      <td>92198968</td>\n",
       "      <td>92246166</td>\n",
       "      <td>-</td>\n",
       "      <td>KRIT1</td>\n",
       "      <td>47198</td>\n",
       "      <td>23266.858150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000002834_ENST00000585841</th>\n",
       "      <td>38869959</td>\n",
       "      <td>38870142</td>\n",
       "      <td>ENSG00000002834</td>\n",
       "      <td>17</td>\n",
       "      <td>gene</td>\n",
       "      <td>38869858</td>\n",
       "      <td>38921770</td>\n",
       "      <td>+</td>\n",
       "      <td>LASP1</td>\n",
       "      <td>51912</td>\n",
       "      <td>29300.639149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285043_newTSS</th>\n",
       "      <td>30064056</td>\n",
       "      <td>30064226</td>\n",
       "      <td>ENSG00000285043</td>\n",
       "      <td>16</td>\n",
       "      <td>gene</td>\n",
       "      <td>30053089</td>\n",
       "      <td>30070420</td>\n",
       "      <td>+</td>\n",
       "      <td>AC093512.2</td>\n",
       "      <td>17331</td>\n",
       "      <td>22023.363957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285331_ENST00000648603</th>\n",
       "      <td>56918242</td>\n",
       "      <td>56918390</td>\n",
       "      <td>ENSG00000285331</td>\n",
       "      <td>15</td>\n",
       "      <td>gene</td>\n",
       "      <td>56795459</td>\n",
       "      <td>56918499</td>\n",
       "      <td>-</td>\n",
       "      <td>AC090517.5</td>\n",
       "      <td>123040</td>\n",
       "      <td>7929.171224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285646_ENST00000650448</th>\n",
       "      <td>11908143</td>\n",
       "      <td>11908221</td>\n",
       "      <td>ENSG00000285646</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>11907939</td>\n",
       "      <td>11914298</td>\n",
       "      <td>+</td>\n",
       "      <td>AL021155.5</td>\n",
       "      <td>6359</td>\n",
       "      <td>13139.479124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285756_newTSS</th>\n",
       "      <td>3882165</td>\n",
       "      <td>3882287</td>\n",
       "      <td>ENSG00000285756</td>\n",
       "      <td>X</td>\n",
       "      <td>gene</td>\n",
       "      <td>3817527</td>\n",
       "      <td>3937855</td>\n",
       "      <td>-</td>\n",
       "      <td>BX890604.2</td>\n",
       "      <td>120328</td>\n",
       "      <td>6140.318460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000286340_ENST00000669651</th>\n",
       "      <td>79078545</td>\n",
       "      <td>79078653</td>\n",
       "      <td>ENSG00000286340</td>\n",
       "      <td>6</td>\n",
       "      <td>gene</td>\n",
       "      <td>79077840</td>\n",
       "      <td>79081371</td>\n",
       "      <td>+</td>\n",
       "      <td>AL356776.2</td>\n",
       "      <td>3531</td>\n",
       "      <td>1454.454024</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2356 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 TSS_start    TSS_end          gene_id  \\\n",
       "transcript_id                                                            \n",
       "ENSG00000000460_ENST00000359326  169794982  169795207  ENSG00000000460   \n",
       "ENSG00000000938_ENST00000457296   27626061   27626162  ENSG00000000938   \n",
       "ENSG00000001626_newTSS           117479966  117480057  ENSG00000001626   \n",
       "ENSG00000001631_ENST00000412043   92245654   92245871  ENSG00000001631   \n",
       "ENSG00000002834_ENST00000585841   38869959   38870142  ENSG00000002834   \n",
       "...                                    ...        ...              ...   \n",
       "ENSG00000285043_newTSS            30064056   30064226  ENSG00000285043   \n",
       "ENSG00000285331_ENST00000648603   56918242   56918390  ENSG00000285331   \n",
       "ENSG00000285646_ENST00000650448   11908143   11908221  ENSG00000285646   \n",
       "ENSG00000285756_newTSS             3882165    3882287  ENSG00000285756   \n",
       "ENSG00000286340_ENST00000669651   79078545   79078653  ENSG00000286340   \n",
       "\n",
       "                                Chromosome Feature      Start        End  \\\n",
       "transcript_id                                                              \n",
       "ENSG00000000460_ENST00000359326          1    gene  169662006  169854080   \n",
       "ENSG00000000938_ENST00000457296          1    gene   27612063   27635185   \n",
       "ENSG00000001626_newTSS                   7    gene  117465783  117715971   \n",
       "ENSG00000001631_ENST00000412043          7    gene   92198968   92246166   \n",
       "ENSG00000002834_ENST00000585841         17    gene   38869858   38921770   \n",
       "...                                    ...     ...        ...        ...   \n",
       "ENSG00000285043_newTSS                  16    gene   30053089   30070420   \n",
       "ENSG00000285331_ENST00000648603         15    gene   56795459   56918499   \n",
       "ENSG00000285646_ENST00000650448          1    gene   11907939   11914298   \n",
       "ENSG00000285756_newTSS                   X    gene    3817527    3937855   \n",
       "ENSG00000286340_ENST00000669651          6    gene   79077840   79081371   \n",
       "\n",
       "                                Strand   gene_name     len         count  \n",
       "transcript_id                                                             \n",
       "ENSG00000000460_ENST00000359326      +    C1orf112  192074  21252.589071  \n",
       "ENSG00000000938_ENST00000457296      -         FGR   23122    618.203238  \n",
       "ENSG00000001626_newTSS               +        CFTR  250188    796.037284  \n",
       "ENSG00000001631_ENST00000412043      -       KRIT1   47198  23266.858150  \n",
       "ENSG00000002834_ENST00000585841      +       LASP1   51912  29300.639149  \n",
       "...                                ...         ...     ...           ...  \n",
       "ENSG00000285043_newTSS               +  AC093512.2   17331  22023.363957  \n",
       "ENSG00000285331_ENST00000648603      -  AC090517.5  123040   7929.171224  \n",
       "ENSG00000285646_ENST00000650448      +  AL021155.5    6359  13139.479124  \n",
       "ENSG00000285756_newTSS               -  BX890604.2  120328   6140.318460  \n",
       "ENSG00000286340_ENST00000669651      +  AL356776.2    3531   1454.454024  \n",
       "\n",
       "[2356 rows x 11 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seconddf=keepdf[1::2]\n",
    "seconddf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "188b277e-6054-42e2-bf7b-720b2e98db3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 37559 × 2356\n",
       "    obs: 'index', 'cell_id', 'batch'\n",
       "    var: 'gene_ids', 'feature_types', 'gene_name'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keepexpadata=expadata[adata.obs.index.tolist(),seconddf['gene_id'].tolist()]\n",
    "keepexpadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "44b8002f-f691-460a-9cc4-b1ef166084b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 37559 × 2356\n",
       "    obs: 'cell_id', 'Anno_level_1', 'Anno_level_2', 'Anno_level_3', 'Anno_level_4', 'Anno_level_5', 'Anno_level_fig1', 'Sample', 'donor', 'organ', 'sort', 'file', 'Anno_stage', 'Age', 'Gender', 'Source'\n",
       "    var: 'gene_ids', 'feature_types', 'gene_name'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keepexpadata.obs=adata.obs\n",
    "keepexpadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f728c135-191c-4f9b-bd0b-c5dab61adb6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 37559 × 2356\n",
       "    obs: 'cell_id', 'Anno_level_1', 'Anno_level_2', 'Anno_level_3', 'Anno_level_4', 'Anno_level_5', 'Anno_level_fig1', 'Sample', 'donor', 'organ', 'sort', 'file', 'Anno_stage', 'Age', 'Gender', 'Source'\n",
       "    var: 'TSS_start', 'TSS_end', 'gene_id', 'Chromosome', 'Feature', 'Start', 'End', 'Strand', 'gene_name', 'len'\n",
       "    uns: 'log1p'\n",
       "    obsm: 'X_umap'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isoform1adata=adata[:,firstdf.index.tolist()]\n",
    "isoform1adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8b5ac8df-90d6-46f2-9b84-ca7421598927",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 37559 × 2356\n",
       "    obs: 'cell_id', 'Anno_level_1', 'Anno_level_2', 'Anno_level_3', 'Anno_level_4', 'Anno_level_5', 'Anno_level_fig1', 'Sample', 'donor', 'organ', 'sort', 'file', 'Anno_stage', 'Age', 'Gender', 'Source'\n",
       "    var: 'TSS_start', 'TSS_end', 'gene_id', 'Chromosome', 'Feature', 'Start', 'End', 'Strand', 'gene_name', 'len'\n",
       "    uns: 'log1p'\n",
       "    obsm: 'X_umap'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isoform2adata=adata[:,seconddf.index.tolist()]\n",
    "isoform2adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b5651261-bca5-4b42-8e40-51942af0fc7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ENSG00000000460_ENST00000359326', 'ENSG00000000938_ENST00000457296',\n",
       "       'ENSG00000001626_newTSS', 'ENSG00000001631_ENST00000412043',\n",
       "       'ENSG00000002834_ENST00000585841', 'ENSG00000002919_ENST00000578861',\n",
       "       'ENSG00000002933_ENST00000004103', 'ENSG00000003402_ENST00000443227',\n",
       "       'ENSG00000004534_ENST00000443081', 'ENSG00000004897_ENST00000531206',\n",
       "       ...\n",
       "       'ENSG00000278558_ENST00000618859', 'ENSG00000280594_newTSS',\n",
       "       'ENSG00000283103_ENST00000658331', 'ENSG00000284634_ENST00000641608',\n",
       "       'ENSG00000284770_ENST00000642463', 'ENSG00000285043_newTSS',\n",
       "       'ENSG00000285331_ENST00000648603', 'ENSG00000285646_ENST00000650448',\n",
       "       'ENSG00000285756_newTSS', 'ENSG00000286340_ENST00000669651'],\n",
       "      dtype='object', name='transcript_id', length=2356)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isoform2adata.var.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "584e89cc-fe96-4d8b-b44b-efd0dbd759d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "keepexpadata.layers['isoform1']=isoform1adata.layers['counts']\n",
    "keepexpadata.layers['isoform2']=isoform2adata.layers['counts']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d574c3e1-3561-4e2c-b106-ffc7e9fcfbe2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 37559 × 2356\n",
       "    obs: 'cell_id', 'Anno_level_1', 'Anno_level_2', 'Anno_level_3', 'Anno_level_4', 'Anno_level_5', 'Anno_level_fig1', 'Sample', 'donor', 'organ', 'sort', 'file', 'Anno_stage', 'Age', 'Gender', 'Source'\n",
       "    var: 'gene_ids', 'feature_types', 'gene_name', 'isoform1_name', 'isoform2_name'\n",
       "    layers: 'isoform1', 'isoform2'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keepexpadata.var['isoform1_name']=isoform1adata.var.index.values\n",
    "keepexpadata.var['isoform2_name']=isoform2adata.var.index.values\n",
    "keepexpadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6595864e-ddaa-4ba3-a6ce-3cd5c356bc22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 37559 × 2356\n",
       "    obs: 'cell_id', 'Anno_level_1', 'Anno_level_2', 'Anno_level_3', 'Anno_level_4', 'Anno_level_5', 'Anno_level_fig1', 'Sample', 'donor', 'organ', 'sort', 'file', 'Anno_stage', 'Age', 'Gender', 'Source'\n",
       "    var: 'gene_ids', 'feature_types', 'gene_name', 'isoform1_name', 'isoform2_name'\n",
       "    layers: 'isoform1', 'isoform2', 'ambiguous'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keepexpadata.layers['ambiguous']=np.zeros((37559, 2356))\n",
    "keepexpadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ebcdc4f6-893e-4428-b14d-96e07b69e5c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "keepexpadata.X=keepexpadata.X.astype('float32')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d2dcad65-7ba7-4a24-b46a-027fdbfa3c54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 37559 × 2356\n",
       "    obs: 'cell_id', 'Anno_level_1', 'Anno_level_2', 'Anno_level_3', 'Anno_level_4', 'Anno_level_5', 'Anno_level_fig1', 'Sample', 'donor', 'organ', 'sort', 'file', 'Anno_stage', 'Age', 'Gender', 'Source'\n",
       "    var: 'gene_ids', 'feature_types', 'gene_name', 'isoform1_name', 'isoform2_name'\n",
       "    layers: 'isoform1', 'isoform2', 'ambiguous'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keepexpadata.layers['isoform1']=keepexpadata.layers['isoform1'].astype('float32')\n",
    "keepexpadata.layers['isoform2']=keepexpadata.layers['isoform2'].astype('float32')\n",
    "keepexpadata.layers['ambiguous']=keepexpadata.layers['ambiguous'].astype('float32')\n",
    "\n",
    "keepexpadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c7b5ab91-8163-451d-b6e5-232555bc591a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 1., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       ...,\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 1., 0., 0.],\n",
       "       [0., 0., 0., ..., 1., 0., 0.]], dtype=float32)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keepexpadata.X.toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9de0c0af-973e-454c-89cb-9332e5703b5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "keepexpadata.write('/storage3/yhhuang/users/ruiyan/development_data/Brie/express_to_brie.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a90bcccf-969b-47f6-9b5b-97b4f027dec4",
   "metadata": {},
   "source": [
    "### Get h5ad file of cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4d778c38-09ac-4fd5-aefe-c8266dd6edb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5e21d2f9-52b3-423c-a7a2-fbea3b12375b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 51001 × 1784\n",
       "    obs: 'cluster', 'UMAP_1', 'UMAP_2', 'patient_ID', 'condition'\n",
       "    var: 'gene_ids', 'feature_types', 'gene_name', 'isoform1_name', 'isoform2_name'\n",
       "    layers: 'ambiguous', 'isoform1', 'isoform2'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata=sc.read('/storage/yhhuang/users/ruiyan/NPC/express_to_brie.h5ad')\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ea6f469c-c35e-44c9-9e8e-dac2b58f4817",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['T cells', 'B cells', 'Myeloids', 'NK', 'Epithelial', 'Fibroblast']\n",
       "Categories (6, object): ['B cells', 'Epithelial', 'Fibroblast', 'Myeloids', 'NK', 'T cells']"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['cluster'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "cbaad4c0-1ac8-4570-893a-0c8b4e6a42bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_celltype_adata(cellType):\n",
    "    celltypeadata=adata[adata.obs['cluster']==cellType,:]\n",
    "    celltype=cellType.replace(' ','_')\n",
    "    file=celltype+'.h5ad'\n",
    "    outputpath=os.path.join('/storage/yhhuang/users/ruiyan/NPC/diffcluster/',file)\n",
    "    celltypeadata.write(outputpath)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4585707b-2147-40a7-93f2-169ed2cb0add",
   "metadata": {},
   "outputs": [],
   "source": [
    "for celltype in adata.obs['cluster'].unique():\n",
    "    get_celltype_adata(celltype)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0105f90-1c4d-43bd-a5a4-af49e53088b9",
   "metadata": {},
   "source": [
    "### get cdr file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "384ed4e9-1163-49b7-9798-15e08c567925",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c1bed97d-e0a3-4628-823e-08ae5c38426d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 51001 × 1784\n",
       "    obs: 'cluster', 'UMAP_1', 'UMAP_2', 'patient_ID', 'condition'\n",
       "    var: 'gene_ids', 'feature_types', 'gene_name', 'isoform1_name', 'isoform2_name'\n",
       "    layers: 'ambiguous', 'isoform1', 'isoform2'"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata=sc.read('/storage/yhhuang/users/ruiyan/NPC/express_to_brie.h5ad')\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "879e651a-da5d-49c5-8016-1c7aa6d4a1a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cdr_file(cellcluster):\n",
    "    cellcluster=cellcluster.replace(' ','_')\n",
    "    inputfile='/storage/yhhuang/users/ruiyan/NPC/diffcluster/'+str(cellcluster)+'.h5ad'\n",
    "    clusteradata=sc.read(inputfile)\n",
    "    diseasedf=pd.DataFrame(clusteradata.obs['condition'])\n",
    "    cdrdiseasedf = pd.get_dummies(diseasedf.condition, prefix='disease')\n",
    "    \n",
    "    cdr=np.array((clusteradata.X > 0).mean(1))[:,0]\n",
    "    cdrdiseasedf['detect_rate']=cdr\n",
    "    cdrdiseasedf.reset_index(inplace=True)\n",
    "    \n",
    "\n",
    "    cdrdiseasedf.drop(labels=['disease_NLH'],axis=1,inplace=True)\n",
    "    print(cdrdiseasedf)\n",
    "    outfile='/storage/yhhuang/users/ruiyan/NPC/diffcluster/NPC_cdr_'+cellcluster+'.tsv'\n",
    "    cdrdiseasedf.to_csv(outfile,sep='\\t',index=None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "8bbd92d5-afec-46e2-aebb-2f7641473939",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     index  disease_NPC  detect_rate\n",
      "0      CACCACTAGGAGTCTG-11            0     0.174888\n",
      "1       GACTAACGTCGGCTCA-9            1     0.135090\n",
      "2       GCAGCCAGTAGCCTCG-5            1     0.043722\n",
      "3       CACAAACGTCAGGACA-2            1     0.143498\n",
      "4       TACGGTATCTCAACTT-1            1     0.108744\n",
      "...                    ...          ...          ...\n",
      "24566   CTAATGGGTGACTACT-1            1     0.147422\n",
      "24567   CAAGAAAAGGCACATG-3            1     0.038117\n",
      "24568  CACATTTGTGCTCTTC-10            1     0.186099\n",
      "24569  CATGGCGAGAGCAATT-10            1     0.172646\n",
      "24570   CTCGAGGGTTCATGGT-5            1     0.154148\n",
      "\n",
      "[24571 rows x 3 columns]\n",
      "                     index  disease_NPC  detect_rate\n",
      "0       GCATGTAGTCTAGTCA-4            1     0.103139\n",
      "1       TAAGAGAAGGAGTCTG-5            1     0.092489\n",
      "2       CCGTACTCAGACAAAT-4            1     0.093610\n",
      "3       TTGCCGTTCAACCATG-4            1     0.086323\n",
      "4      GAGTCCGCATCGGTTA-12            0     0.246637\n",
      "...                    ...          ...          ...\n",
      "24291  TGGTTAGTCCGTAGGC-11            0     0.115471\n",
      "24292   CCTTTCTGTACTTAGC-9            1     0.182735\n",
      "24293   CAGCTGGTCGCCTGAG-4            1     0.125000\n",
      "24294   CGGTTAACAGGCGATA-9            1     0.227578\n",
      "24295  CGTAGCGCAAAGGAAG-11            0     0.071188\n",
      "\n",
      "[24296 rows x 3 columns]\n",
      "                    index  disease_NPC  detect_rate\n",
      "0      GCCAAATCATCTCCCA-3            1     0.188341\n",
      "1      CATTATCGTTTGCATG-3            1     0.142377\n",
      "2      TAGCCGGGTCAGAATA-3            1     0.207399\n",
      "3      GTGCGGTCATTTCACT-3            1     0.192825\n",
      "4      CACATTTAGCCTTGAT-2            1     0.147422\n",
      "...                   ...          ...          ...\n",
      "1076   CATGGCGTCCGCAGTG-3            1     0.332399\n",
      "1077   GTCATTTCAGCCAATT-3            1     0.315583\n",
      "1078   CGGAGTCAGCTGCGAA-9            1     0.263453\n",
      "1079   TAAGCGTCAAAGGTGC-2            1     0.204036\n",
      "1080  GTCACAACATGTCGAT-12            0     0.252803\n",
      "\n",
      "[1081 rows x 3 columns]\n",
      "                   index  disease_NPC  detect_rate\n",
      "0     TGACGGCTCTCGGACG-9            1     0.156390\n",
      "1     CATTATCTCGATCCCT-2            1     0.162556\n",
      "2     GGCGACTAGGATATAC-9            1     0.329036\n",
      "3     GTGCAGCAGGGCATGT-9            1     0.136771\n",
      "4    AGGTCATCACTTACGA-13            0     0.225336\n",
      "..                   ...          ...          ...\n",
      "868  GTCTCGTAGAGACTAT-11            0     0.124439\n",
      "869  AACTCTTTCATACGGT-12            0     0.113229\n",
      "870  CCTCAGTAGCAATCTC-10            1     0.095852\n",
      "871   TACTTACAGATCCCAT-2            1     0.142377\n",
      "872  GTATCTTGTAAATGAC-12            0     0.158072\n",
      "\n",
      "[873 rows x 3 columns]\n",
      "                  index  disease_NPC  detect_rate\n",
      "0    CTTACCGCATATGAGA-5            1     0.198430\n",
      "1    CTTAACTAGACTTTCG-1            1     0.047646\n",
      "2    TGCCCTACACCAGCAC-3            1     0.272422\n",
      "3    GGATTACGTAGCTCCG-3            1     0.208520\n",
      "4    GCGCCAATCTCTGAGA-5            1     0.131726\n",
      "..                  ...          ...          ...\n",
      "131  GATTCAGGTACTTAGC-1            1     0.089686\n",
      "132  TCGCGAGGTATTCTCT-2            1     0.020740\n",
      "133  TCAGGTAGTCGTCTTC-9            1     0.248318\n",
      "134  AGATCTGGTTAGGGTG-5            1     0.056614\n",
      "135  CTCGTCACATAGACTC-1            1     0.082960\n",
      "\n",
      "[136 rows x 3 columns]\n",
      "                  index  disease_NPC  detect_rate\n",
      "0    CCTTACGGTACAGCAG-1            1     0.215807\n",
      "1    CGTTAGACAAGTCTGT-3            1     0.040359\n",
      "2    GACGGCTGTAAGTGTA-9            1     0.334641\n",
      "3    GTATCTTTCATATCGG-9            1     0.161996\n",
      "4    GTAACGTCAGTCAGCC-9            1     0.285874\n",
      "5    CAGCAGCTCACAGGCC-1            1     0.249439\n",
      "6   ACATCAGAGGAGCGTT-11            0     0.149103\n",
      "7    AGAGCTTCATATGGTC-1            1     0.106502\n",
      "8    TACTTGTAGCTACCTA-1            1     0.198430\n",
      "9    ATAACGCGTACAGTTC-1            1     0.135650\n",
      "10   TAAGTGCAGGTGTTAA-1            1     0.171525\n",
      "11  GACGGCTTCAGTTCGA-11            0     0.144619\n",
      "12  GTTCATTGTGCGAAAC-11            0     0.172085\n",
      "13   CTAGAGTCAGTCAGAG-3            1     0.093610\n",
      "14  CGTTAGAAGCTATGCT-11            0     0.196749\n",
      "15   AACTCAGCACGGCTAC-3            1     0.059978\n",
      "16   TCTGGAAAGCTCCTCT-1            1     0.209081\n",
      "17   CCTACACAGAGACTTA-3            1     0.252242\n",
      "18   ACGGCCAAGGGATGGG-1            1     0.047646\n",
      "19   CGTCACTCACATGGGA-1            1     0.167601\n",
      "20   CTGCGGACAAGCCTAT-9            1     0.260090\n",
      "21   TAGTTGGCATTAGGCT-1            1     0.237668\n",
      "22   GTCCTCAGTAGGACAC-3            1     0.088004\n",
      "23   AACTTTCCATTGGCGC-1            1     0.026345\n",
      "24   TGGGCGTCAAGTACCT-3            1     0.064462\n",
      "25  TCTTCGGAGATCACGG-11            0     0.257287\n",
      "26   CCTAGCTTCGAGGTAG-1            1     0.189462\n",
      "27   ACTTGTTCATGCAATC-9            1     0.048206\n",
      "28  TATCTCAAGCGTAATA-11            0     0.274664\n",
      "29   TGCGCAGAGGAGTACC-1            1     0.200673\n",
      "30   TAGACCACATGGGACA-1            1     0.338565\n",
      "31   AACCATGTCAGCAACT-1            1     0.034193\n",
      "32   GCACATAAGCTGTTCA-1            1     0.161996\n",
      "33   CCGGGATAGACTAGGC-1            1     0.213565\n",
      "34   CGAGCACCAGCTGTTA-3            1     0.280830\n",
      "35   CGTAGGCAGAGCCTAG-3            1     0.071749\n",
      "36   TCTCATAGTCAAGCGA-1            1     0.145740\n",
      "37   CCGGGATCAGTGAGTG-1            1     0.239910\n",
      "38   TTAGTTCGTAGGACAC-1            1     0.026906\n",
      "39   TTCTCCTAGCCAGTTT-1            1     0.321749\n",
      "40   ACTTACTCATACGCTA-1            1     0.205717\n",
      "41   CTACCCAAGTGCTGCC-3            1     0.088004\n",
      "42   CGAACATCATGTCTCC-1            1     0.155830\n",
      "43   CTCGAGGTCAGATAAG-2            1     0.298767\n"
     ]
    }
   ],
   "source": [
    "for celltype in adata.obs['cluster'].unique():\n",
    "    get_cdr_file(celltype)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9822d7b5-ba14-4bfe-b7be-aeefe2bcf894",
   "metadata": {},
   "source": [
    "## Run Brie2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6587c94f-0189-49f7-8c63-5e63659127a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "arr=(NK Myeloids Epithelial Fibroblast T_cells B_cells)\n",
    "for i in ${arr[@]}\n",
    "do\n",
    "echao $i\n",
    "brie-quant -i /storage/yhhuang/users/ruiyan/NPC/diffcluster/${i}.h5ad   -c /storage/yhhuang/users/ruiyan/NPC/diffcluster/NPC_cdr_${i}.tsv   -o /storage/yhhuang/users/ruiyan/NPC/diffcluster/Brie_out/brie_NPC_${i}.h5ad --batchSize 1000000 --minCell 10  --interceptMode gene --testBase full --LRTindex 0\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e9a690f-c460-4cf1-a839-d3953e2b3959",
   "metadata": {},
   "source": [
    "# Do filtering (standard: FDR<0.01)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b71eb6cd-9765-4977-a3f2-cd76d33bc3b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "62386968-8f91-418b-9d20-2016994b146a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sign_gene(celltype):\n",
    "    celltype=celltype.replace(' ','_')\n",
    "    inputfile='/storage/yhhuang/users/ruiyan/NPC/diffcluster/Brie_out/brie_NPC_'+celltype+'.brie_ident.tsv'\n",
    "    briedf=pd.read_csv(inputfile,delimiter='\\t')\n",
    "    print(briedf)\n",
    "    signTdf=briedf[briedf['disease_NPC_FDR']<0.01]\n",
    "    outputfile='/storage/yhhuang/users/ruiyan/NPC/diffcluster/sign_Gene/'+celltype+'.csv'\n",
    "    signTdf.to_csv(outputfile,index=None)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "f5c31caa-c05c-49b5-ac43-9ffaba7de82b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 51001 × 4226\n",
       "    obs: 'cluster', 'UMAP_1', 'UMAP_2', 'patient_ID', 'condition'\n",
       "    var: 'TSS_start', 'TSS_end', 'gene_id', 'Chromosome', 'Feature', 'Start', 'End', 'Strand', 'gene_name', 'len'\n",
       "    uns: 'log1p'\n",
       "    obsm: 'X_umap'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata=sc.read('/storage/yhhuang/users/ruiyan/NPC/srafiledir/scTSS_out/NPC_new_normalized.h5ad')\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a681fcae-990b-4096-9257-b29892f8c850",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['T cells', 'B cells', 'Myeloids', 'NK', 'Epithelial', 'Fibroblast']\n",
       "Categories (6, object): ['B cells', 'Epithelial', 'Fibroblast', 'Myeloids', 'NK', 'T cells']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['cluster'].unique()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "37b42618-97d5-4da2-b2eb-a2a0e7878feb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "               GeneID  n_counts  n_counts_uniq      cdr  intercept   sigma  \\\n",
      "0     ENSG00000000460       737            737  0.03345    -1.1680  2.4840   \n",
      "1     ENSG00000000938       178            178  0.01604    -1.1560  1.7140   \n",
      "2     ENSG00000001167       896            896  0.04786    -2.3100  1.3030   \n",
      "3     ENSG00000001460        63             63  0.01701    -3.0630  1.6360   \n",
      "4     ENSG00000001631      1919           1919  0.12460    -3.1440  0.7132   \n",
      "...               ...       ...            ...      ...        ...     ...   \n",
      "1623  ENSG00000278558       225            225  0.01457    -1.4710  1.9890   \n",
      "1624  ENSG00000278709       325            325  0.01404    -1.6940  1.2780   \n",
      "1625  ENSG00000284691        58             58  0.07920     0.7147  2.7520   \n",
      "1626  ENSG00000285077       248            248  0.01095    -1.0360  1.2130   \n",
      "1627  ENSG00000285106      2254           2254  0.25070    -3.4440  0.4510   \n",
      "\n",
      "      disease_NPC_ceoff  disease_NPC_ELBO_gain  disease_NPC_pval  \\\n",
      "0                2.4160                14.7300      5.680000e-08   \n",
      "1                0.1911                -0.1767      1.000000e+00   \n",
      "2               -0.7912                 2.8350      1.726000e-02   \n",
      "3                2.5510                 3.8490      5.529000e-03   \n",
      "4                0.5392                 2.0660      4.209000e-02   \n",
      "...                 ...                    ...               ...   \n",
      "1623             0.6478                 0.7807      2.115000e-01   \n",
      "1624            -0.1897                 0.4706      3.320000e-01   \n",
      "1625            -3.2030                 2.4280      2.756000e-02   \n",
      "1626            -1.7810                 5.7940      6.640000e-04   \n",
      "1627             0.1752                 0.3850      3.802000e-01   \n",
      "\n",
      "      disease_NPC_FDR  \n",
      "0        3.881000e-07  \n",
      "1        1.000000e+00  \n",
      "2        3.930000e-02  \n",
      "3        1.417000e-02  \n",
      "4        7.845000e-02  \n",
      "...               ...  \n",
      "1623     3.607000e-01  \n",
      "1624     4.814000e-01  \n",
      "1625     1.998000e-01  \n",
      "1626     9.627000e-03  \n",
      "1627     5.251000e-01  \n",
      "\n",
      "[1628 rows x 10 columns]\n",
      "               GeneID  n_counts  n_counts_uniq       cdr  intercept   sigma  \\\n",
      "0     ENSG00000000460       406            406  0.021810    -1.7500  1.8650   \n",
      "1     ENSG00000000938       648            648  0.070180    -1.6300  1.5620   \n",
      "2     ENSG00000001167      1159           1159  0.056430    -3.0280  0.5059   \n",
      "3     ENSG00000001460        51             51  0.019100    -1.7020  1.6740   \n",
      "4     ENSG00000001631      2249           2249  0.140300    -3.4540  0.4162   \n",
      "...               ...       ...            ...       ...        ...     ...   \n",
      "1686  ENSG00000278558       199            199  0.017410    -0.9191  1.4620   \n",
      "1687  ENSG00000278709       208            208  0.009261    -2.6830  1.1840   \n",
      "1688  ENSG00000284691       132            132  0.102500    -0.7034  1.9560   \n",
      "1689  ENSG00000285077       209            209  0.010820    -1.1990  1.3750   \n",
      "1690  ENSG00000285106      4588           4588  0.345900    -4.1340  0.3500   \n",
      "\n",
      "      disease_NPC_ceoff  disease_NPC_ELBO_gain  disease_NPC_pval  \\\n",
      "0              -0.25090               -0.07494          1.000000   \n",
      "1              -0.72930                4.60100          0.002418   \n",
      "2              -0.54070                1.60500          0.073180   \n",
      "3               1.82100                2.21200          0.035420   \n",
      "4               0.36860                2.08600          0.041080   \n",
      "...                 ...                    ...               ...   \n",
      "1686           -0.56610                0.58170          0.280800   \n",
      "1687           -0.32390               -0.03098          1.000000   \n",
      "1688           -0.31870               -0.35440          1.000000   \n",
      "1689           -0.87820                1.56200          0.077100   \n",
      "1690            0.06543                0.14450          0.590800   \n",
      "\n",
      "      disease_NPC_FDR  \n",
      "0             1.00000  \n",
      "1             0.01016  \n",
      "2             0.14640  \n",
      "3             0.08922  \n",
      "4             0.09080  \n",
      "...               ...  \n",
      "1686          0.77210  \n",
      "1687          1.00000  \n",
      "1688          1.00000  \n",
      "1689          0.28270  \n",
      "1690          1.00000  \n",
      "\n",
      "[1691 rows x 10 columns]\n",
      "              GeneID  n_counts  n_counts_uniq     cdr  intercept   sigma  \\\n",
      "0    ENSG00000000938        56             56  0.6679     -2.809  0.5546   \n",
      "1    ENSG00000001631        74             74  0.1471     -3.409  0.3655   \n",
      "2    ENSG00000002919        85             85  0.1508     -2.483  0.2853   \n",
      "3    ENSG00000002933      3347           3347  0.5171     -3.202  0.7587   \n",
      "4    ENSG00000003402       231            231  0.6827     -3.494  0.8103   \n",
      "..               ...       ...            ...     ...        ...     ...   \n",
      "880  ENSG00000272449       322            322  0.2146     -3.746  0.9700   \n",
      "881  ENSG00000275052       169            169  0.2877     -3.268  0.2276   \n",
      "882  ENSG00000275302       720            720  0.3867     -3.919  0.1885   \n",
      "883  ENSG00000278311       137            137  0.3293     -3.106  0.5179   \n",
      "884  ENSG00000285106        80             80  0.1554     -3.554  0.4217   \n",
      "\n",
      "     disease_NPC_ceoff  disease_NPC_ELBO_gain  disease_NPC_pval  \\\n",
      "0             -0.08966               -0.05511           1.00000   \n",
      "1              0.94970                0.12620           0.61540   \n",
      "2             -1.30500                0.64450           0.25620   \n",
      "3             -0.49830                1.06300           0.14490   \n",
      "4              0.20600               -0.00956           1.00000   \n",
      "..                 ...                    ...               ...   \n",
      "880           -0.80330                0.73430           0.22560   \n",
      "881            0.15660                0.04677           0.75970   \n",
      "882           -1.08200                0.67740           0.24450   \n",
      "883            0.73260                1.41900           0.09202   \n",
      "884            1.49600                0.50260           0.31610   \n",
      "\n",
      "     disease_NPC_FDR  \n",
      "0             1.0000  \n",
      "1             0.8972  \n",
      "2             0.7373  \n",
      "3             0.6969  \n",
      "4             1.0000  \n",
      "..               ...  \n",
      "880           0.7370  \n",
      "881           0.9844  \n",
      "882           0.7373  \n",
      "883           0.6216  \n",
      "884           0.7851  \n",
      "\n",
      "[885 rows x 10 columns]\n",
      "              GeneID  n_counts  n_counts_uniq      cdr  intercept   sigma  \\\n",
      "0    ENSG00000000938       175            175  0.32990    -2.5880  2.0220   \n",
      "1    ENSG00000001631        69             69  0.12830    -3.9130  0.3209   \n",
      "2    ENSG00000003402        94             94  0.38830    -1.5070  0.3469   \n",
      "3    ENSG00000004478        58             58  0.07102    -4.4580  0.3612   \n",
      "4    ENSG00000005302        55             55  0.09851     2.6360  1.0190   \n",
      "..               ...       ...            ...      ...        ...     ...   \n",
      "665  ENSG00000275052       127            127  0.22910    -3.9200  0.2428   \n",
      "666  ENSG00000275302       451            451  0.42500    -4.2520  0.5549   \n",
      "667  ENSG00000276070       195            195  0.18100    -4.9070  0.2960   \n",
      "668  ENSG00000278311       139            139  0.32420    -0.6992  0.7648   \n",
      "669  ENSG00000285106        63             63  0.20160    -2.7690  0.4947   \n",
      "\n",
      "     disease_NPC_ceoff  disease_NPC_ELBO_gain  disease_NPC_pval  \\\n",
      "0              0.36360                0.62420            0.2639   \n",
      "1              0.72120                0.13970            0.5971   \n",
      "2              0.70660                0.62610            0.2631   \n",
      "3              0.02145               -0.11980            1.0000   \n",
      "4             -0.86530                0.60940            0.2696   \n",
      "..                 ...                    ...               ...   \n",
      "665            1.21300                0.64090            0.2576   \n",
      "666            0.12970                0.22640            0.5010   \n",
      "667            0.37080               -0.01399            1.0000   \n",
      "668           -0.26870                0.07555            0.6975   \n",
      "669           -0.19390               -0.02064            1.0000   \n",
      "\n",
      "     disease_NPC_FDR  \n",
      "0             0.8728  \n",
      "1             0.8728  \n",
      "2             0.8728  \n",
      "3             1.0000  \n",
      "4             0.8728  \n",
      "..               ...  \n",
      "665           0.8728  \n",
      "666           0.8728  \n",
      "667           1.0000  \n",
      "668           0.9441  \n",
      "669           1.0000  \n",
      "\n",
      "[670 rows x 10 columns]\n",
      "             GeneID  n_counts  n_counts_uniq     cdr  intercept    sigma  \\\n",
      "0   ENSG00000004478        98             98  0.4485  -3.544000  0.22890   \n",
      "1   ENSG00000025708        60             60  0.6618  -0.517300  1.54400   \n",
      "2   ENSG00000037241        84             84  0.3750  -8.162000  1.67000   \n",
      "3   ENSG00000065057        76             76  0.2868   3.322000  4.98600   \n",
      "4   ENSG00000069849       122            122  0.4632  -5.296000  0.19320   \n",
      "5   ENSG00000072778        66             66  0.4338  -2.085000  0.25500   \n",
      "6   ENSG00000073803        70             70  0.3676  -3.638000  0.66070   \n",
      "7   ENSG00000088986        67             67  0.7721  -4.810000  0.27690   \n",
      "8   ENSG00000093010       114            114  0.5147   0.191200  0.98780   \n",
      "9   ENSG00000095066        54             54  0.3750   0.475500  0.26090   \n",
      "10  ENSG00000100097       141            141  0.5147  -0.935000  0.34210   \n",
      "11  ENSG00000103490        94             94  0.6103  -4.591000  0.21900   \n",
      "12  ENSG00000103769        81             81  0.6250  -2.402000  0.14810   \n",
      "13  ENSG00000104763        58             58  0.3971  -0.987300  0.30790   \n",
      "14  ENSG00000104870        61             61  0.4338  -1.926000  0.33920   \n",
      "15  ENSG00000110492       422            422  0.7574  -4.472000  0.22020   \n",
      "16  ENSG00000111639        83             83  0.6103  -5.377000  0.17650   \n",
      "17  ENSG00000112096        50             50  0.5294  -0.070120  1.08000   \n",
      "18  ENSG00000114331        51             51  0.4485  -2.086000  0.20450   \n",
      "19  ENSG00000116044        64             64  0.5074  -6.058000  0.46190   \n",
      "20  ENSG00000116171        59             59  0.5368  -4.291000  0.22210   \n",
      "21  ENSG00000119917        78             78  0.3456  -2.130000  0.84490   \n",
      "22  ENSG00000122378       112            112  0.4485  -0.005322  0.97620   \n",
      "23  ENSG00000127838        72             72  0.4118  -3.130000  0.29340   \n",
      "24  ENSG00000128951        52             52  0.4853  -0.988100  0.58310   \n",
      "25  ENSG00000130770        60             60  0.6471  -4.507000  0.21390   \n",
      "26  ENSG00000132386       131            131  0.3015   5.882000  0.23720   \n",
      "27  ENSG00000135046       286            286  0.5441  -4.380000  0.09177   \n",
      "28  ENSG00000135480       264            264  0.2721   3.982000  4.33000   \n",
      "29  ENSG00000135821       486            486  0.7721  -5.229000  0.14420   \n",
      "30  ENSG00000136997        82             82  0.3824  -7.875000  0.27650   \n",
      "31  ENSG00000137312        53             53  0.3824  -5.400000  0.26150   \n",
      "32  ENSG00000148120        53             53  0.4118   4.403000  0.83940   \n",
      "33  ENSG00000149131        60             60  0.3015   0.788600  1.59100   \n",
      "34  ENSG00000149925       112            112  0.8676  -0.583000  0.18640   \n",
      "35  ENSG00000158092       130            130  0.4853  -0.333500  1.01600   \n",
      "36  ENSG00000161011        75             75  0.7574  -5.714000  0.17930   \n",
      "37  ENSG00000170035        79             79  0.4926  -4.061000  0.21580   \n",
      "38  ENSG00000170421       667            667  0.7868  -3.232000  1.38200   \n",
      "39  ENSG00000170892        59             59  0.3824   2.322000  0.22330   \n",
      "40  ENSG00000171401       598            598  0.3015   0.963200  3.06600   \n",
      "41  ENSG00000172366        52             52  0.2941   1.783000  0.20430   \n",
      "42  ENSG00000173848       128            128  0.5368  -3.648000  0.19630   \n",
      "43  ENSG00000175334        54             54  0.6103  -1.812000  0.26990   \n",
      "44  ENSG00000176209        53             53  0.3603  -4.105000  0.24790   \n",
      "45  ENSG00000177426        65             65  0.3897   0.707000  0.35510   \n",
      "46  ENSG00000181885       110            110  0.4559   5.726000  0.26210   \n",
      "47  ENSG00000186184        66             66  0.5809  -6.932000  0.23820   \n",
      "48  ENSG00000196465        80             80  0.4338  -2.784000  0.26330   \n",
      "49  ENSG00000198355        65             65  0.4779  -4.807000  0.28330   \n",
      "50  ENSG00000204580        67             67  0.5000   5.152000  1.19100   \n",
      "51  ENSG00000214253        59             59  0.6176  -1.219000  0.19440   \n",
      "52  ENSG00000272398        92             92  0.4926  -2.291000  0.19860   \n",
      "\n",
      "    disease_NPC_ceoff  disease_NPC_ELBO_gain  disease_NPC_pval  \\\n",
      "0            -0.36940               -0.02302          1.000000   \n",
      "1             0.37360                0.15970          0.572000   \n",
      "2             3.29300                1.11200          0.135900   \n",
      "3            -6.08400                0.61820          0.266100   \n",
      "4             0.71070                0.16310          0.567800   \n",
      "5             0.69470               -0.01443          1.000000   \n",
      "6            -0.46320                0.06411          0.720300   \n",
      "7             1.51800                0.16150          0.569800   \n",
      "8             0.29840               -0.46040          1.000000   \n",
      "9            -1.20400                0.31790          0.425200   \n",
      "10           -2.18800                0.01321          0.870900   \n",
      "11            1.05100                0.09645          0.660500   \n",
      "12           -2.21700                0.82540          0.198800   \n",
      "13           -2.23000                0.71900          0.230500   \n",
      "14            0.84300                0.42500          0.356600   \n",
      "15           -0.09718                0.03923          0.779400   \n",
      "16            1.49200                0.28540          0.449900   \n",
      "17           -0.64620                0.06374          0.721100   \n",
      "18            0.85200               -0.02547          1.000000   \n",
      "19            3.58300                1.79800          0.057900   \n",
      "20            0.79860                0.32770          0.418200   \n",
      "21           -1.31600                0.64720          0.255200   \n",
      "22            0.42290                0.18060          0.547800   \n",
      "23           -0.07422                0.04054          0.775800   \n",
      "24            0.28730                0.20240          0.524700   \n",
      "25            0.85440                0.05345          0.743700   \n",
      "26           -1.63200                0.34570          0.405700   \n",
      "27            1.22800                1.06500          0.144500   \n",
      "28           -2.40700                0.22010          0.507000   \n",
      "29            0.57560                0.12560          0.616200   \n",
      "30            1.57100                0.30410          0.435500   \n",
      "31            1.31800                0.07684          0.695000   \n",
      "32           -2.81600                0.42430          0.357000   \n",
      "33           -0.04626               -0.40770          1.000000   \n",
      "34           -0.59020                0.14430          0.591200   \n",
      "35           -0.06989               -0.01768          1.000000   \n",
      "36            1.46900                0.28200          0.452600   \n",
      "37            2.72700                1.11900          0.134600   \n",
      "38            0.02907               -0.14380          1.000000   \n",
      "39           -1.97600                2.74800          0.019060   \n",
      "40           -6.04900                5.93900          0.000568   \n",
      "41           -3.40600                2.54300          0.024110   \n",
      "42           -1.31100                0.35330          0.400500   \n",
      "43            2.18900                0.16560          0.565000   \n",
      "44            0.94210                0.01415          0.866400   \n",
      "45           -0.85900                0.24300          0.485700   \n",
      "46           -0.55010                0.02916          0.809200   \n",
      "47            1.58300                0.37600          0.385800   \n",
      "48           -0.27990                0.12480          0.617400   \n",
      "49            1.60700                0.40970          0.365300   \n",
      "50           -4.23800                2.25500          0.033680   \n",
      "51           -2.78000                1.12300          0.133900   \n",
      "52           -0.90490                0.29230          0.444500   \n",
      "\n",
      "    disease_NPC_FDR  \n",
      "0           1.00000  \n",
      "1           0.90890  \n",
      "2           0.85070  \n",
      "3           0.90890  \n",
      "4           0.90890  \n",
      "5           1.00000  \n",
      "6           0.95540  \n",
      "7           0.90890  \n",
      "8           1.00000  \n",
      "9           0.90890  \n",
      "10          1.00000  \n",
      "11          0.94610  \n",
      "12          0.90890  \n",
      "13          0.90890  \n",
      "14          0.90890  \n",
      "15          0.96060  \n",
      "16          0.90890  \n",
      "17          0.95540  \n",
      "18          1.00000  \n",
      "19          0.61370  \n",
      "20          0.90890  \n",
      "21          0.90890  \n",
      "22          0.90890  \n",
      "23          0.96060  \n",
      "24          0.90890  \n",
      "25          0.96060  \n",
      "26          0.90890  \n",
      "27          0.85070  \n",
      "28          0.90890  \n",
      "29          0.90890  \n",
      "30          0.90890  \n",
      "31          0.95540  \n",
      "32          0.90890  \n",
      "33          1.00000  \n",
      "34          0.90890  \n",
      "35          1.00000  \n",
      "36          0.90890  \n",
      "37          0.85070  \n",
      "38          1.00000  \n",
      "39          0.42590  \n",
      "40          0.03011  \n",
      "41          0.42590  \n",
      "42          0.90890  \n",
      "43          0.90890  \n",
      "44          1.00000  \n",
      "45          0.90890  \n",
      "46          0.97470  \n",
      "47          0.90890  \n",
      "48          0.90890  \n",
      "49          0.90890  \n",
      "50          0.44630  \n",
      "51          0.85070  \n",
      "52          0.90890  \n",
      "            GeneID  n_counts  n_counts_uniq     cdr  intercept    sigma  \\\n",
      "0  ENSG00000100097        84             84  0.5909   -6.19000  0.98510   \n",
      "1  ENSG00000105974       170            170  0.7500    6.49000  1.33900   \n",
      "2  ENSG00000122786       183            183  0.6364    6.64000  0.11090   \n",
      "3  ENSG00000140416       161            161  0.4545    9.00000  3.85600   \n",
      "4  ENSG00000144837        55             55  0.2273    0.02651  0.07303   \n",
      "5  ENSG00000149131       165            165  0.7727   -1.14300  0.08982   \n",
      "6  ENSG00000173848        60             60  0.3864   -3.50500  2.03800   \n",
      "\n",
      "   disease_NPC_ceoff  disease_NPC_ELBO_gain  disease_NPC_pval  disease_NPC_FDR  \n",
      "0              2.048               0.170800           0.55890          0.78250  \n",
      "1              1.087               0.807400           0.20380          0.47560  \n",
      "2             -1.231              -0.029020           1.00000          1.00000  \n",
      "3             -6.168               0.593000           0.27610          0.48320  \n",
      "4              2.597              -0.008677           1.00000          1.00000  \n",
      "5              1.097               3.092000           0.01288          0.09019  \n",
      "6             -6.037               2.191000           0.03630          0.12710  \n"
     ]
    }
   ],
   "source": [
    "for celltype in adata.obs['cluster'].unique():\n",
    "    get_sign_gene(celltype)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3cb6b1a-0a29-46d0-b44a-832d369f485b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "newPy3",
   "language": "python",
   "name": "newpy3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
